File: s3/moves.ml
open Str
open Printf
open Scanf
open Queue

let read_input filename =
  let ic = open_in filename in
  let n = int_of_string (input_line ic) in
  let grid = Array.make_matrix n n 0 in
  for i = 0 to n - 1 do
    let line = input_line ic in
    let nums = Str.split (Str.regexp " +") line in
    List.iteri (fun j num -> grid.(i).(j) <- int_of_string num) nums
  done;
  close_in ic;
  (n, grid)

let directions = [|
  (-1, 0); (1, 0); (0, -1); (0, 1);
  (-1, -1); (-1, 1); (1, -1); (1, 1)
|]

let in_bounds n x y = x >= 0 && x < n && y >= 0 && y < n

let bfs n grid =
  let start = (0, 0) in
  let goal = (n-1, n-1) in
  let visited = Array.make_matrix n n false in
  let queue = create () in
  let path = Hashtbl.create 100 in

  add (start, []) queue;
  visited.(0).(0) <- true;

  let rec search () =
    if is_empty queue then None
    else
      let ((x, y), moves) = take queue in
      if (x, y) = goal then Some (List.rev moves)
      else
        let current_cars = grid.(x).(y) in
        Array.iter (fun (dx, dy) ->
          let nx, ny = x + dx, y + dy in
          if in_bounds n nx ny && not visited.(nx).(ny) && grid.(nx).(ny) < current_cars then
            let direction =
              match dx, dy with
              | -1, 0 -> "N"
              | 1, 0 -> "S"
              | 0, -1 -> "W"
              | 0, 1 -> "E"
              | -1, -1 -> "NW"
              | -1, 1 -> "NE"
              | 1, -1 -> "SW"
              | 1, 1 -> "SE"
              | _ -> failwith "Invalid direction"
            in
            visited.(nx).(ny) <- true;
            add ((nx, ny), direction :: moves) queue;
        ) directions;
        search ()
  in
  search ()

  let () =
  let (n, grid) = read_input Sys.argv.(1) in
  match bfs n grid with
  | Some moves -> 
    let moves_string = "[" ^ (String.concat "," moves) ^ "]" in
    printf "%s" moves_string
  | None -> printf "IMPOSSIBLE"

File: s3/moves.py
import sys
import heapq as heapriorityqueue

DIRECTIONS = {
    'N': (-1, 0), 'S': (1, 0), 'E': (0, 1), 'W': (0, -1),
    'NE': (-1, 1), 'NW': (-1, -1), 'SE': (1, 1), 'SW': (1, -1)
}

def parse_input(file_path):
    with open(file_path, 'r') as file:
        lines = file.readlines()
    n = int(lines[0].strip())
    grid = [list(map(int, line.split())) for line in lines[1:]]
    return n, grid

def is_valid(x, y, n):
    return 0 <= x < n and 0 <= y < n

def find_path(n, grid):
    start = (0, 0)
    goal = (n - 1, n - 1)
    priorityqueue = [(0, start, [])]
    visited = set()
    
    while priorityqueue:
        cost, (x, y), path = heapriorityqueue.heappop(priorityqueue)
        
        if (x, y) in visited:
            continue
        visited.add((x, y))
        
        if (x, y) == goal:
            return path
        
        for direction, (dx, dy) in DIRECTIONS.items():
            nx, ny = x + dx, y + dy
            if is_valid(nx, ny, n) and grid[nx][ny] < grid[x][y]:
                heapriorityqueue.heappush(priorityqueue, (cost + 1, (nx, ny), path + [direction]))
    
    return "IMPOSSIBLE"

def main():
    if len(sys.argv) != 2:
        print("Usage: python3 moves.py <input_file>")
        return

    input_file = sys.argv[1]
    n, grid = parse_input(input_file)
    result = find_path(n, grid)
    if result == "IMPOSSIBLE":
        print(result)
    else:
        print('[' + ','.join(result) + ']')

if __name__ == "__main__":
    main()